src/coreclr/vm/jitinterface.cpp

```diff
@@ -4221,7 +4237,14 @@ TypeCompareState CEEInfo::compareTypesForEquality(
     // If neither type is a canonical subtype, type handle comparison suffices
     if (!hnd1.IsCanonicalSubtype() && !hnd2.IsCanonicalSubtype())
     {
-        result = (hnd1 == hnd2 ? TypeCompareState::Must : TypeCompareState::MustNot);
+        if (hnd1.IsFnPtrType() || hnd2.IsFnPtrType())
+        {
+            result = TypeCompareState::May;
```

Comment:
The goal here is not to embed this aspect of type system policy in the jit.

If the runtime changes how it represents function pointer runtime types to create a broader 1-1 mapping between runtime handles and runtime types, this will still be correct, but could be updated to allow the jit to optimize more broady too.

Reply:
The way I see it is that this is changing the runtime to compensate for an incorrect assumption the JIT makes (injectivity of `TypeHandleToRuntimeType`), instead of fixing the wrong JIT assumption. I suppose that really depends on what you consider by type equality though. But taken to the extreme, a hypothetical runtime that did not support managed reflection at all could return `null` from the helper always, but that wouldn't mean all types may be the same.

Regardless I believe we will be fixing this `typeof` oddity soon, so I don't see a point in a more complicated fix.

Reply:
I would leave the function pointer cases broken and not complicate the code for them. If you tried, I think you would find more places than are broken in addition to what is touched in this PR.

If we wanted to fix the function pointer cases for .NET 7, I would rather look into bringing over small subset of changes from #11354 - to have 1:1 mapping between System.RuntimeType and runtime/JIT typehandles.

cc @steveharter

Reply:
@jakobbotsch I'm not sure what you are proposing as an alternative. The jit has very limited ability to reason about class handles. So inevitably it has to ask the runtime for help with something like this. And the jit only assumes 1-1 mapping when the runtime says it's ok to make that assumption.

The ability of the jit to eagerly fold type comparisons is crucial for pruning all the type specialization that now happens in the BCL. So the jit can't just stop doing this all together and can't (on its own) figure out when it can safely do this optimization. Hence the current design...

Reply:
What I suggested in my comment above is what I was thinking – i.e. introducing a new helper (such as `compareRuntimeTypesForEquality`, that would still take class handles). However given that `canInlineTypeCheck` exists that seems appropriate too.

But in that case I am unsure why `compareTypesForEquality` needed to be changed, wouldn't the change in `canInlineTypeCheck` suffice if the VN folding also consulted it?